### ISSUE TITLE
remove public dependency on `ignore`

### ISSUE THREAD

**Issue by BurntSushi** [Gilnaa/globwalk#15]
It looking at the API of this crate, I noticed that an [`ignore::Error`](https://docs.rs/globwalk/0.3.1/globwalk/type.GlobError.html) is re-exported through a type alias. I think we should probably try to avoid that, since the `ignore` crate isn't something I'd consider stable yet.

This crate also has a public dependency on the walkdir crate, but this one is a bit trickier to remove. Namely, `WalkError` exposes walkdir's error type, but the iterator itself also exposes walkdir's `DirEntry` type. This public dependency might be OK though, since I largely consider walkdir's API stable and there are no current plans for releasing any new major versions.

**Comment by Gilnaa** [comment]
What is the preferred mode of operation? 
Copying the error details to a local type or just destroying the evidence?

**Comment by BurntSushi** [comment]
@Gilnaa Good question. There are a few points in the design space, but if it were me, I'd probably do this:

```rust
#[derive(Clone, Debug)]
pub struct GlobError(ignore::Error);

impl From<std::io::Error> for GlobError { ... }

// Should convert to the underlying I/O error if possible,
// and otherwise generate an io::ErrorKind::Other error.
impl From<GlobError> for std::io::Error { ... }

impl std::error::Error for GlobError { ... }

impl fmt::Display for GlobError { ... }

impl GlobError {
    fn io_error(&self) -> Option<&std::io::Error> { ... }
    fn depth(&self) -> Option<usize> { self.0.depth() }
}
```

And that should just about do it. I would honestly probably try to coalesce the errors emitted by this crate into one error type if possible to keep things simple.

**Comment by epage** [comment]
I originally didn't combine them since they served different end-user purposes (config vs runtime error) but I agree a single error is nicer. If we want to keep the distinction, we could always expose a Kind.

**Comment by Gilnaa** [comment]
Closed by f16cf37
